includes/subscriptions/class-wc-payments-subscriptions-onboarding-handler.php

```diff
@@ -0,0 +1,193 @@
+<?php
+/**
+ * Class WC_Payments_Subscriptions_Onboarding_Handler
+ *
+ * @package WooCommerce\Payments
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * A class to handle the onboarding of subscriptions products. Created subscription products will be set to draft until
+ *   onboarding is completed.
+ */
+class WC_Payments_Subscriptions_Onboarding_Handler {
+
+	/**
+	 * Transient to store subscription publish blocked error.
+	 *
+	 * @const string
+	 */
+	const WCPAY_SUBSCRIPTION_PUBLISH_BLOCKED_NOTICE = 'wcpay_subscription_onboarding_blocked_notice';
+
+	/**
+	 * Option for holding an array of product id's to publish post onboarding.
+	 *
+	 * @const string
+	 */
+	const WCPAY_SUBSCRIPTION_AUTO_PUBLISH_PRODUCTS = 'wcpay_subscription_onboarding_products';
+
+	/**
+	 * Transient to store notice that subscriptions were auto published.
+	 *
+	 * @const string
+	 */
+	const WCPAY_SUBSCRIPTION_AUTO_PUBLISH_NOTICE = 'wcpay_subscription_onboarding_published_notice';
+
+	/**
+	 * The account service instance.
+	 *
+	 * @var WC_Payments_Account|null
+	 */
+	private static $account = null;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		add_filter( 'publish_product', [ $this, 'product_published' ], 10, 3 );
```

Comment:
We should also use `add_action` here rather than `add_filter`.

`do_action` should be hooked with `add_action` and `apply_filters` should be hooked with `add_filter`.

This hook in particular uses `do_action`, ref: https://developer.wordpress.org/reference/hooks/new_status_post-post_type/

Reply:
I missed this, was playing around with other actions and was trying one that had multiple arguments. Will update.